route_following_plugin/src/route_following_plugin.cpp

```diff
@@ -92,139 +103,105 @@ namespace route_following_plugin
+        while(current_progress < total_maneuver_length && i < maneuvers.size()){
+           double maneuver_end_dist = GET_MANEUVER_PROPERTY(maneuvers[i],end_dist);
+           current_progress += maneuver_end_dist;
+           //Update time progress for maneuvers
+           resp.new_plan.maneuvers.push_back(maneuvers[i]);
+           i++;
         }
-        resp.new_plan.maneuvers.push_back(
-            composeStopandWaitManeuverMessage(current_progress,total_maneuver_length,
-            speed_progress,shortest_path[last_lanelet_index].id(),
-            shortest_path[last_lanelet_index].id(),time_progress,time_req_to_stop));
         
-        ROS_DEBUG_STREAM("Done Loop: approaching_route_end: " << approaching_route_end);
-        if(approaching_route_end){
-            ros::Time end_time = ros::Time::now();
-            if (resp.new_plan.maneuvers.size() > 0) {
-                end_time = resp.new_plan.maneuvers.back().lane_following_maneuver.end_time;
-            }
-            
-            resp.new_plan.maneuvers.push_back(
-                composeStopandWaitManeuverMessage(current_progress,total_maneuver_length,
-                speed_progress,shortest_path[last_lanelet_index].id(),
-                shortest_path[last_lanelet_index].id(),end_time,time_req_to_stop)); 
-        }
         if(resp.new_plan.maneuvers.size() == 0)
         {
             ROS_WARN_STREAM("Cannot plan maneuver because no route is found");
         }
         return true;
     }
-    void RouteFollowingPlugin::pose_cb(const geometry_msgs::PoseStampedConstPtr& msg)
-    {
-        pose_msg_ = geometry_msgs::PoseStamped(*msg.get());
+
+    void RouteFollowingPlugin::set_maneuver_start_dist(cav_msgs::Maneuver& maneuver, double start_dist){
+        switch(maneuver.type){
+            case cav_msgs::Maneuver::LANE_FOLLOWING :
+                maneuver.lane_following_maneuver.start_dist = start_dist;
+            case cav_msgs::Maneuver::LANE_CHANGE :
+                maneuver.lane_change_maneuver.start_dist = start_dist;
+            case cav_msgs::Maneuver::INTERSECTION_TRANSIT_STRAIGHT :
+                maneuver.intersection_transit_straight_maneuver.start_dist = start_dist;
+            case cav_msgs::Maneuver::INTERSECTION_TRANSIT_LEFT_TURN :
+                maneuver.intersection_transit_left_turn_maneuver.start_dist = start_dist;
+            case cav_msgs::Maneuver::INTERSECTION_TRANSIT_RIGHT_TURN :
+                maneuver.intersection_transit_right_turn_maneuver.start_dist = start_dist;
+            case cav_msgs::Maneuver::STOP_AND_WAIT :
+                maneuver.stop_and_wait_maneuver.start_dist = start_dist;
+            default :
+                ROS_DEBUG_STREAM("Invalid maneuver type");
+        }
+
     }
-    void RouteFollowingPlugin::twist_cd(const geometry_msgs::TwistStampedConstPtr& msg)
-    {
-        current_speed_ = msg->twist.linear.x;
+    
+    double RouteFollowingPlugin::get_final_speed(cav_msgs::Maneuver maneuver){
```

Comment:
This one is a member function because not all maneuvers have target_speed as a parameter, for stop and wait the end speed is always 0.0, its not a part of the message parameters

Reply:
Could the macro not be updated to account for that instead of a separate function? If it needs to be separate from the macro, a free function would still be preferable to a member function I'd think. While it's used by the RFP, it still doesn't really have any real relationship to it at a logical level that would necessitate this function belonging to the RFP.

Reply:
This function is not being used anymore